tst_IrcConnection::testSasl crashes on armv7hl

It used to work fine but started crashing when we tried to rebuild it after Mageia got Qt 5.15.2, however this may not be the source of the problem as many other things changed since last time it was rebuilt 9 months earlier.

```
QWARN  : tst_IrcConnection::testSasl() QMetaObject::newInstance: type IrcCapabilityMessage does not inherit QObject

=== Received signal at function time: 1ms, total time: 134ms, dumping stack ===
=== End of stack trace ===
QFATAL : tst_IrcConnection::testSasl() Received signal 11
         Function time: 1ms Total time: 134ms
FAIL!  : tst_IrcConnection::testSasl() Received a fatal error.
   Loc: [Unknown file(0)]
Totals: 848 passed, 1 failed, 0 skipped, 0 blacklisted, 142ms
********* Finished testing of tst_IrcConnection *********
make[3]: *** [Makefile:299: check] Aborted (core dumped)
make[3]: Leaving directory '/home/iurt/rpmbuild/BUILD/libcommuni-3.5.0/tests/auto/ircconnection'
```

--- a/include/IrcModel/ircbuffer.h
+++ b/include/IrcModel/ircbuffer.h
@@ -89,6 +89,8 @@ public:
 
     Q_INVOKABLE bool sendCommand(IrcCommand* command);
 
+    virtual IrcBuffer *clone(QObject* parent = nullptr);
+
 public Q_SLOTS:
     void setName(const QString& name);
     void setPrefix(const QString& prefix);

--- a/include/IrcModel/ircchannel.h
+++ b/include/IrcModel/ircchannel.h
@@ -54,6 +54,8 @@ public:
 
     bool isActive() const override;
 
+    IrcBuffer *clone(QObject* parent = nullptr) override;
+
 public Q_SLOTS:
     void who();
     void join(const QString& key = QString());

--- a/src/core/ircmessage.cpp
+++ b/src/core/ircmessage.cpp
@@ -242,40 +242,41 @@ IRC_BEGIN_NAMESPACE
     \brief The message is an implicit "reply" after joining a channel.
  */
 
-static const QMetaObject* irc_command_meta_object(const QString& command)
+static IrcMessage* irc_create_message(const QString& command, IrcConnection* connection)
 {
-    static QHash<QString, const QMetaObject*> metaObjects;
-    if (metaObjects.isEmpty()) {
-        metaObjects.insert("ACCOUNT", &IrcAccountMessage::staticMetaObject);
-        metaObjects.insert("AWAY", &IrcAwayMessage::staticMetaObject);
-        metaObjects.insert("BATCH", &IrcBatchMessage::staticMetaObject);
-        metaObjects.insert("CAP", &IrcCapabilityMessage::staticMetaObject);
-        metaObjects.insert("ERROR", &IrcErrorMessage::staticMetaObject);
-        metaObjects.insert("CHGHOST", &IrcHostChangeMessage::staticMetaObject);
-        metaObjects.insert("INVITE", &IrcInviteMessage::staticMetaObject);
-        metaObjects.insert("JOIN", &IrcJoinMessage::staticMetaObject);
-        metaObjects.insert("KICK", &IrcKickMessage::staticMetaObject);
-        metaObjects.insert("MODE", &IrcModeMessage::staticMetaObject);
-        metaObjects.insert("NICK", &IrcNickMessage::staticMetaObject);
-        metaObjects.insert("NOTICE", &IrcNoticeMessage::staticMetaObject);
-        metaObjects.insert("PART", &IrcPartMessage::staticMetaObject);
-        metaObjects.insert("PING", &IrcPingMessage::staticMetaObject);
-        metaObjects.insert("PONG", &IrcPongMessage::staticMetaObject);
-        metaObjects.insert("PRIVMSG", &IrcPrivateMessage::staticMetaObject);
-        metaObjects.insert("QUIT", &IrcQuitMessage::staticMetaObject);
-        metaObjects.insert("TOPIC", &IrcTopicMessage::staticMetaObject);
-    }
+    typedef std::function<IrcMessage *(IrcConnection *)> IrcMessageFactory;
 
-    const QMetaObject* metaObject = metaObjects.value(command.toUpper());
-    if (!metaObject) {
-        bool ok = false;
-        command.toInt(&ok);
-        if (ok)
-            metaObject = &IrcNumericMessage::staticMetaObject;
-    }
-    if (!metaObject)
-        metaObject = &IrcMessage::staticMetaObject;
-    return metaObject;
+    static const QHash<QString, IrcMessageFactory> factories = {
+        {"ACCOUNT", [](IrcConnection* c) { return new IrcAccountMessage(c); }},
+        {"AWAY", [](IrcConnection* c) { return new IrcAwayMessage(c); }},
+        {"BATCH", [](IrcConnection* c) { return new IrcBatchMessage(c); }},
+        {"CAP", [](IrcConnection* c) { return new IrcCapabilityMessage(c); }},
+        {"ERROR", [](IrcConnection* c) { return new IrcErrorMessage(c); }},
+        {"CHGHOST", [](IrcConnection* c) { return new IrcHostChangeMessage(c); }},
+        {"INVITE", [](IrcConnection* c) { return new IrcInviteMessage(c); }},
+        {"JOIN", [](IrcConnection* c) { return new IrcJoinMessage(c); }},
+        {"KICK", [](IrcConnection* c) { return new IrcKickMessage(c); }},
+        {"MODE", [](IrcConnection* c) { return new IrcModeMessage(c); }},
+        {"NICK", [](IrcConnection* c) { return new IrcNickMessage(c); }},
+        {"NOTICE", [](IrcConnection* c) { return new IrcNoticeMessage(c); }},
+        {"PART", [](IrcConnection* c) { return new IrcPartMessage(c); }},
+        {"PING", [](IrcConnection* c) { return new IrcPingMessage(c); }},
+        {"PONG", [](IrcConnection* c) { return new IrcPongMessage(c); }},
+        {"PRIVMSG", [](IrcConnection* c) { return new IrcPrivateMessage(c); }},
+        {"QUIT", [](IrcConnection* c) { return new IrcQuitMessage(c); }},
+        {"TOPIC", [](IrcConnection* c) { return new IrcTopicMessage(c); }},
+    };
+
+    IrcMessageFactory factory = factories.value(command);
+    if (factory)
+        return factory(connection);
+
+    bool isNumeric = false;
+    int number = command.toInt(&isNumeric);
+    if (isNumeric && number > 0)
+        return new IrcNumericMessage(connection);
+
+    return new IrcMessage(connection);
 }
 
 /*!
@@ -690,19 +691,15 @@ void IrcMessage::setTag(const QString& name, const QVariant& value)
  */
 IrcMessage* IrcMessage::fromData(const QByteArray& data, IrcConnection* connection)
 {
-    IrcMessage* message = nullptr;
     IrcMessageData md = IrcMessageData::fromData(data);
-    const QMetaObject* metaObject = irc_command_meta_object(md.command);
-    if (metaObject) {
-        message = qobject_cast<IrcMessage*>(metaObject->newInstance(Q_ARG(IrcConnection*, connection)));
-        Q_ASSERT(message);
-        message->d_ptr->data = md;
-        QByteArray tag = md.tags.value("time");
-        if (!tag.isEmpty()) {
-            QDateTime ts = QDateTime::fromString(QString::fromUtf8(tag), Qt::ISODate);
-            if (ts.isValid())
-                message->d_ptr->timeStamp = ts.toTimeSpec(Qt::LocalTime);
-        }
+    IrcMessage* message = irc_create_message(md.command, connection);
+    Q_ASSERT(message);
+    message->d_ptr->data = md;
+    QByteArray tag = md.tags.value("time");
+    if (!tag.isEmpty()) {
+        QDateTime ts = QDateTime::fromString(QString::fromUtf8(tag), Qt::ISODate);
+        if (ts.isValid())
+            message->d_ptr->timeStamp = ts.toTimeSpec(Qt::LocalTime);
     }
     return message;
 }
@@ -712,15 +709,11 @@ IrcMessage* IrcMessage::fromData(const QByteArray& data, IrcConnection* connecti
  */
 IrcMessage* IrcMessage::fromParameters(const QString& prefix, const QString& command, const QStringList& parameters, IrcConnection* connection)
 {
-    IrcMessage* message = nullptr;
-    const QMetaObject* metaObject = irc_command_meta_object(command);
-    if (metaObject) {
-        message = qobject_cast<IrcMessage*>(metaObject->newInstance(Q_ARG(IrcConnection*, connection)));
-        Q_ASSERT(message);
-        message->setPrefix(prefix);
-        message->setCommand(command);
-        message->setParameters(parameters);
-    }
+    IrcMessage* message = irc_create_message(command, connection);
+    Q_ASSERT(message);
+    message->setPrefix(prefix);
+    message->setCommand(command);
+    message->setParameters(parameters);
     return message;
 }
 
@@ -732,7 +725,7 @@ IrcMessage* IrcMessage::fromParameters(const QString& prefix, const QString& com
 IrcMessage* IrcMessage::clone(QObject* parent) const
 {
     Q_D(const IrcMessage);
-    IrcMessage* msg = qobject_cast<IrcMessage*>(metaObject()->newInstance(Q_ARG(IrcConnection*, d->connection)));
+    IrcMessage* msg = irc_create_message(d->command(), d->connection);
     if (msg) {
         msg->setParent(parent);
         IrcMessagePrivate* p = IrcMessagePrivate::get(msg);

--- a/src/model/ircbuffer.cpp
+++ b/src/model/ircbuffer.cpp
@@ -582,6 +582,16 @@ bool IrcBuffer::sendCommand(IrcCommand* command)
 }
 
 /*!
+    \since 3.7
+
+    Clones the buffer with an optional \a parent.
+ */
+IrcBuffer *IrcBuffer::clone(QObject *parent)
+{
+    return new IrcBuffer(parent);
+}
+
+/*!
     Emits messageReceived() with \a message.
 
     IrcBufferModel handles only buffer specific messages and delivers them

--- a/src/model/ircbuffermodel.cpp
+++ b/src/model/ircbuffermodel.cpp
@@ -1104,8 +1104,7 @@ IrcBuffer* IrcBufferModel::createBuffer(const QString& title)
 {
     Q_D(IrcBufferModel);
     Q_UNUSED(title);
-    QObject* instance = d->bufferProto->metaObject()->newInstance(Q_ARG(QObject*, this));
-    return qobject_cast<IrcBuffer*>(instance);
+    return d->bufferProto->clone(this);
 }
 
 /*!
@@ -1123,8 +1122,7 @@ IrcChannel* IrcBufferModel::createChannel(const QString& title)
 {
     Q_D(IrcBufferModel);
     Q_UNUSED(title);
-    QObject* instance = d->channelProto->metaObject()->newInstance(Q_ARG(QObject*, this));
-    return qobject_cast<IrcChannel*>(instance);
+    return qobject_cast<IrcChannel*>(d->channelProto->clone(this));
 }
 
 /*!
@@ -1251,7 +1249,7 @@ QModelIndex IrcBufferModel::index(int row, int column, const QModelIndex& parent
 
     The prototype is used by the default implementation of createBuffer().
 
-    \note The prototype must have an invokable constructor.
+    \note A custom buffer prototype must reimplement \ref IrcBuffer::clone().
 
     \par Access functions:
     \li \ref IrcBuffer* <b>bufferPrototype</b>() const
@@ -1279,7 +1277,7 @@ void IrcBufferModel::setBufferPrototype(IrcBuffer* prototype)
 
     The prototype is used by the default implementation of createChannel().
 
-    \note The prototype must have an invokable constructor.
+    \note A custom channel prototype must reimplement \ref IrcChannel::clone().
 
     \par Access functions:
     \li \ref IrcChannel* <b>channelPrototype</b>() const

--- a/src/model/ircchannel.cpp
+++ b/src/model/ircchannel.cpp
@@ -573,6 +573,11 @@ bool IrcChannel::isActive() const
     return IrcBuffer::isActive() && d->active;
 }
 
+IrcBuffer *IrcChannel::clone(QObject *parent)
+{
+    return new IrcChannel(parent);
+}
+
 /*!
     \since 3.3
 

